### PR TITLE
feat: type-safe useIsSweepNeededQuery

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

Same rationale as https://github.com/shapeshift/web/pull/7144 i.e removing type-narrowing from consumers, and handling it in the hook itself, while removing the ugly, unsafe `queryFn` which was handling optionals at runtime by throwing, in favor of a good ol' type-safe async fn and `skipToken`.

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

- loosely contributes to https://github.com/shapeshift/web/issues/6996

Low - ensure sweep needed checks are still happy across the app (savers, THOR LP)
 
> What protocols, transaction types or contract interactions might be affected by this PR?

## Testing

- Sweep needed checks are still happy in THORChain LP and savers

### Engineering

- ☝🏽 

### Operations

- ☝🏽 

### Savers Deposit

- Sweep needed

https://jam.dev/c/4c2d4938-122e-465d-8f0d-5f7f68276be4

- Sweep not needed

https://jam.dev/c/58b01df9-215e-4ede-a3d9-920f187b599d

### Savers withdraws

- Sweep needed

https://jam.dev/c/c419bac3-9e42-49d6-b25d-44ecebb8dcca

- Sweep not needed

https://jam.dev/c/abb3dca3-0915-484c-8184-5124a97fb621

### THORChain LP

- Sweep needed

https://jam.dev/c/a82d9921-d8af-4bfc-ae9c-1b47a34e28a5

- Sweep not needed

https://jam.dev/c/6ba92bdf-3aa6-4ce4-b31a-6259e86fbf55

### CODE REVIEW

File: src/features/defi/providers/thorchain-savers/components/ThorchainSaversManager/Deposit/components/Deposit.tsx [hunk 1]
```diff
@@ -183,7 +184,7 @@ export const Deposit: React.FC<DepositProps> = ({
     accountId: accountId ?? null,
     amountCryptoBaseUnit: toBaseUnit(inputValues?.cryptoAmount, asset?.precision ?? 0),
     memo: thorchainSaversDepositQuote?.memo ?? null,
-    fromAddress,
+    fromAddress: fromAddress ?? null,
```
Comment (gomesalexandre) [inline src/features/defi/providers/thorchain-savers/components/ThorchainSaversManager/Deposit/components/Deposit.tsx:187]: should also be `T | undefined` in `useSendThorTx` but didn't feel like polluting this diff for this

File: src/features/defi/providers/thorchain-savers/components/ThorchainSaversManager/Deposit/components/Deposit.tsx [hunk 2]
```diff
@@ -660,12 +661,13 @@ export const Deposit: React.FC<DepositProps> = ({
...
 
       const isSweepNeededQueryKey: IsSweepNeededQueryKey = ['isSweepNeeded', isSweepNeededQueryArgs]
-      const _isSweepNeeded = isSweepNeededQueryEnabled
-        ? await queryClient.fetchQuery({
-            queryKey: isSweepNeededQueryKey,
-            queryFn: isSweepNeededQueryFn,
-          })
-        : undefined
+      const _isSweepNeeded =
+        isSweepNeededQueryEnabled && isGetSweepNeededInput(isSweepNeededQueryArgs)
```
Comment (gomesalexandre) [inline src/features/defi/providers/thorchain-savers/components/ThorchainSaversManager/Deposit/components/Deposit.tsx:665]: We didn't have type-safety here and the `queryFn` itself was handling this by throwing, which is wrong - now decoupled from react-query by using `getIsSwepNeeded` async fn directly, which is not accepting partials

File: src/pages/Lending/hooks/useIsSweepNeededQuery.ts [hunk 3]
```diff
@@ -51,37 +48,15 @@ const getIsSweepNeeded = async ({
...
-  if (!address) throw new Error('address is required')
-
-  const isSweepNeeded = await getIsSweepNeeded({
-    assetId,
-    address,
-    amountCryptoBaseUnit,
-    txFeeCryptoBaseUnit,
-  })
-  return isSweepNeeded
-}
```
Comment (gomesalexandre) [inline src/pages/Lending/hooks/useIsSweepNeededQuery.ts:67]: 👋🏽

File: src/pages/Lending/hooks/useIsSweepNeededQuery.ts [hunk 4]
```diff
@@ -51,37 +48,15 @@ const getIsSweepNeeded = async ({
...
-    [assetId, address, amountCryptoBaseUnit, txFeeCryptoBaseUnit],
+    () => ['isSweepNeeded', input],
+    [input],
   )
 
   const isSweepNeededQuery = useQuery({
     queryKey: isSweepNeededQueryKey,
-    queryFn,
-    enabled: Boolean(enabled && address && assetId),
+    queryFn: enabled && isGetSweepNeededInput(input) ? () => getIsSweepNeeded(input) : skipToken,
```
Comment (gomesalexandre) [inline src/pages/Lending/hooks/useIsSweepNeededQuery.ts:59]: Type-narrowing now handled by the query hook itself

**Review by 0xApotheosis — APPROVED** [review]
Does all the things!

**Savers Deposit**
Sweep needed

Screenshot 2024-06-18 at 7 27 46 PM

Sweep not needed

Screenshot 2024-06-18 at 7 28 13 PM

**Savers withdraws**
Sweep needed

Screenshot 2024-06-18 at 7 29 47 PM

Sweep not needed

Screenshot 2024-06-18 at 7 28 50 PM


**THORChain LP deposit**
Sweep needed

Screenshot 2024-06-18 at 7 30 18 PM

Sweep not needed

Screenshot 2024-06-18 at 7 30 45 PM

**THORChain LP withdraw**

Sweep needed

Screenshot 2024-06-18 at 7 31 06 PM

Sweep not needed

Screenshot 2024-06-18 at 7 31 41 PM

**Pull request merged**